projects/dendrites/gaussian_classification/run_dendritic_network.py

```diff
@@ -0,0 +1,344 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+"""
+This module runs a dendritic network in continual learning setting where each task
+consists of learning to classify samples drawn from one of two multivariate normal
+distributions.
+
+Dendritic weights can either be hardcoded or learned. All output heads are used for
+both training and inference.
+
+Usage: adjust the config parameters `weight_init`, `dendrite_init`, and `kw` (all in
+`model_args`) to control the type of forward weight initialization, dendritic weight
+initialization, and k-Winners usage, respectively.
+"""
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+from torch import nn
+
+from nupic.research.frameworks.dendrites import AbsoluteMaxGatingDendriticLayer
+from nupic.research.frameworks.vernon import ContinualLearningExperiment, mixins
+from nupic.torch.duty_cycle_metrics import max_entropy
+from nupic.torch.modules import KWinners, SparseWeights, rezero_weights
+from projects.dendrites.gaussian_classification.gaussian import GaussianDataset
+
+
+def hardcode_dendritic_weights(dendrite_segments, context_vectors):
+    num_units, num_segments, dim_context = dendrite_segments.weights.size()
+    num_contexts, _ = context_vectors.size()
+    fixed_segment_weights = torch.zeros((num_units, num_segments, dim_context))
+
+    for i in range(num_units):
+        context_perm = context_vectors[torch.randperm(num_contexts), :]
+        fixed_segment_weights[i, :, :] = 1.0 * (context_perm[:, :] > 0)
+        fixed_segment_weights[i, 1:, :] = -1.0 * fixed_segment_weights[i, 1:, :]
+        del context_perm
+
+    dendrite_segments.weights.data = fixed_segment_weights
+
+
+# ------ Experiment class
+class DendritesExperiment(mixins.RezeroWeights,
+                          ContinualLearningExperiment):
+    pass
+
+
+# ------ Network
+class DendriticMLP(nn.Module):
+    """
+    Dendrite segments receive context directly as input.
+
+                    _____
+                   |_____|    # classifier layer, no dendrite input
+                      ^
+                      |
+                  _________
+    context -->  |_________|  # second linear layer with dendrites
+                      ^
+                      |
+                  _________
+    context -->  |_________|  # first linear layer with dendrites
+                      ^
+                      |
+                    input
+    """
+
+    def __init__(self, input_size, output_size, hidden_size, num_segments, dim_context,
+                 weight_init, dendrite_init, kw, duty_cycle_period=15000,
```

Comment:
In the batch configurations we'd want `duty_cycle_period` to be the default, not 15,000

Reply:
Fixed.